app/models/composition_saver.rb

```diff
@@ -0,0 +1,150 @@
+class CompositionSaver
+  attr_reader :error_type, :error_value, :composition
+
+  class Error < StandardError; end
+
+  def initialize(user:, session_id:)
+    @user = user
+    @session_id = session_id
+  end
+
+  def save(data)
+    if data[:player_id] || data[:player_name]
+      player = init_player(data)
+      unless player.persisted? && !player.changed? || player.save
+        @error_type = 'player'
+        @error_value = player.errors
+        return
+      end
+    end
+
+    map = if data[:map_segment_id]
+      map_segment = MapSegment.find(data[:map_segment_id])
+      map_segment.map
+    elsif data[:map_id]
+      Map.find(data[:map_id])
+    end
+
+    if data[:composition_id] || map
+      @composition = init_composition(data, map: map)
+      unless @composition.persisted? || @composition.save
+        @error_type = 'composition'
+        @error_value = @composition.errors
+        return
+      end
+    end
+
+    if player && @composition
+      comp_player = init_composition_player(data, player: player,
+                                            composition: @composition)
+      unless comp_player.persisted? || comp_player.save
+        @error_type = 'composition_player'
+        @error_value = comp_player.errors
+        return
+      end
+    end
+
+    if player && map_segment && data[:hero_id]
+      selection = init_player_selection(data, composition: @composition,
+                                        player: player, map_segment: map_segment)
+      unless selection.persisted? && !selection.changed? || selection.save
+        @error_type = 'player_selection'
+        @error_value = selection.errors
+        return
+      end
+    end
+
+    true
+  end
```

Comment:
I feel like this is pretty complex to read. There is a mix of declarative and imperative code here that makes the conditional branching a bit much to load into the brain (or maybe that's just how I feel on a Saturday) 😛 ?

I think one thing that would help in this context is to consider treating `players` as a select list rather than a text field in the team composition react component. Then we could give the user an option for `create player`, which would allow us to take care of player record creation independently of team composition record creation. I think that also would be a nice feature to have regardless for users, so they always have a select list of their previously created `players` when they switch to a new map to create a new composition. What do you think?